src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -468,10 +469,13 @@ protected override void Dispose(bool disposing)
                 _canRead = false;
                 _canWrite = false;
                 _canSeek = false;
-                // Don't set the buffer to null, to avoid a NullReferenceException
-                // when users have a race condition in their code (ie, they call
-                // Close when calling another method on Stream like Read).
-                //_buffer = null;
+
+                if (_buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
```

Comment:
Even with these changes, there is still of course a possibility that misuse of FileStream could result in some corruption.  The check for IsClosed avoids the possibility that the buffer is returned while the OS calls are in progress, but it doesn't address other situations, like a race where thread 1 calls Write and gets a reference to _buffer, then thread 2 calls Dispose and nulls out the buffer and returns it, and then thread 1 copies data into the buffer (without the OS involved at all).  I see a few options:
1. We could simply close this PR.
2. We could keep the changes around CopyToAsync but revert the changes around using ArrayPool for FileStream's buffer.
3. We could address this completely by wrapping the body of all read/write operations in _handle.DangerousAddRef/DangerousRelease calls.  We'd effectively be using the safe handle as a lock such that if any part of any operation is in progress, we simply don't put the buffer back into the pool.  The downside to this of course is the extra calls per operation.
4. We could stay firm on the notion that Dispose'ing a FileStream instance while it's actively being used is simply wrong, a programming error, and while we try to minimize resulting corruption from it, we don't make any guarantees.

I don't think there's any need for (1), as we should at least be able to do (2).  Opinions on (3) and (4)?